rosbag2_cpp/src/rosbag2_cpp/reindexers/sequential_reindexer.cpp

```diff
@@ -0,0 +1,265 @@
+  converter_(nullptr),
+  metadata_io_(std::move(metadata_io)),
+  converter_factory_(std::move(converter_factory))
+{}
+
+SequentialReindexer::~SequentialReindexer()
+{
+  reset();
+}
+
+void SequentialReindexer::reset()
+{
+  if (storage_) {
+    storage_.reset();
+  }
+}
+
+
+bool SequentialReindexer::comp_rel_file(
+  const std::string & first_path, const std::string & second_path)
+{
+  std::regex regex_rule(".*_(\\d+)\\.db3", std::regex_constants::ECMAScript);
+
+  std::smatch first_match;
+  std::smatch second_match;
+
+  auto first_regex_good = std::regex_match(first_path, first_match, regex_rule);
+  auto second_regex_good = std::regex_match(second_path, second_match, regex_rule);
+
+  // Make sure the paths have regex matches
+  if (!first_regex_good || !second_regex_good) {
+    throw std::runtime_error("Malformed relative file name. Expected numerical identifier.");
+  }
+
+  // Convert database numbers to uint
+  u_int32_t first_db_num = std::stoul(first_match.str(1), nullptr, 10);
+  u_int32_t second_db_num = std::stoul(second_match.str(1), nullptr, 10);
+
+  return first_db_num < second_db_num;
+}
+
+std::vector<std::string> SequentialReindexer::get_database_files(const std::string & base_folder)
+{
+  // Look in the uri directory to see what database files are there
+  std::vector<std::string> output;
+  for (auto & p_ : boost::filesystem::directory_iterator(base_folder)) {
+    // We are ONLY interested in database files
+    if (p_.path().extension() != ".db3") {
+      continue;
+    }
+
+    output.emplace_back(p_.path().c_str());
+    std::cout << "Found path: " << p_.path().c_str() << "\n";
+  }
+
+  // Sort relative file path by database number
+  std::sort(
+    output.begin(), output.end(),
+    [](std::string a, std::string b) {return comp_rel_file(a, b);});
+
+  return output;
+}
+
+void SequentialReindexer::open(
+  const std::string & database_file,
+  const StorageOptions & storage_options)
+{
+  // Since this is a reindexing operation, assume that there is no metadata.yaml file.
+  // As such, ask the storage with the given URI for its metadata.
+  storage_ = storage_factory_->open_read_only(
+    database_file, storage_options.storage_id);
+  if (!storage_) {
+    throw std::runtime_error{"No storage could be initialized. Abort"};
+  }
+}
+
+void SequentialReindexer::fill_topics_metadata()
+{
+  rcpputils::check_true(storage_ != nullptr, "Bag is not open. Call open() before reading.");
+  topics_metadata_.clear();
+  topics_metadata_.reserve(metadata_.topics_with_message_count.size());
+  for (const auto & topic_information : metadata_.topics_with_message_count) {
+    topics_metadata_.push_back(topic_information.topic_metadata);
+  }
+}
+
+void SequentialReindexer::init_metadata(const std::vector<std::string> & files)
+{
+  metadata_ = rosbag2_storage::BagMetadata{};
+
+  // This reindexer will only work on SQLite files, so this can't change
+  metadata_.storage_identifier = "sqlite3";
```

Comment:
that has to be more generic. What happens if I try to reindex a ros1 bagfile? Or some custom storage?

Reply:
Yeah, I thought `sequential_reindexer` was supposed to be the reindexer that worked on the `sqlite` files. Not sure why. I guess I'm derpy like that.